### PR TITLE
[Live object detection] Added example project to showcase possibility of live object detection

### PR DESCRIPTION
**Pull request by VadymPinchuk**
Added example project to showcase possibility of live object detection
This project is based on old and not updated versions of live object detection, but uses latest flutter-tflite and other dependencies.

### CODE REVIEW

**Comment by fepfitra** [comment]
@VadymPinchuk I ran example from your fork [here](https://github.com/VadymPinchuk/flutter-tflite.git) with custom model which I got [here](https://colab.research.google.com/github/EdjeElectronics/TensorFlow-Lite-Object-Detection-on-Android-and-Raspberry-Pi/blob/master/Train_TFLite2_Object_Detction_Model.ipynb#scrollTo=5tdDbTmHYwu-) but I got an error
image
Why this happened?

**Comment by VadymPinchuk** [comment]
@fepfitra Thanks for highlighting this. It is just different model which is trained in another way. I am not ML specialist, so I cant assist you here. While for the purpose of example I have used model and input/output configurations from the official sample, which works on a single input. I would suggest to check the model and what kind of input/output is expects.

**Comment by PaulTR** [comment]
Hey, added this to my review list. _hopefully_ will find time tomorrow, otherwise early next week :)

**Comment by VadymPinchuk** [comment]
@PaulTR  this sample is still not perfect. I would love to make it simpler, but don't want to import additional dependencies. Will be happy to receive feedback and improve it :) next week

**Comment by VadymPinchuk** [comment]
@PaulTR  now I am happy of the quality of the code :) Can be reviewed. Thank you

**Comment by PaulTR** [comment]
So a couple things with this:

Did you have a chance to test it with iOS? I'm getting the camera feed on my iPad, but not getting the bounding boxes. It also kind of squishes the video feed.

Would it also be possible to support rotating (at least in Android; iOS rotates but doesn't display)?

Thanks!

**Comment by VadymPinchuk** [comment]
@PaulTR I can try to run on my work iOS device to check how it works. As well - can improve state management for rotation. I am on it

**Comment by VadymPinchuk** [comment]
@PaulTR  I cant run any apps on my work iPhone, so I am not able to test iOS version.
Regarding orientation, I would love to fix it, but till not faced camera rotation issue, which always gives 90 degrees. It will require additional work for covering all the things. If it is legal requirement, I will work on it during the week. Thanks

**Comment by MichaelRinger** [comment]
Thanks for the object detection example. What is the average prediction time running on a single image?

**Comment by VadymPinchuk** [comment]
@MichaelHudgins  depends on the image quality and each specific frame. It jumps between 100 to 300 ms for medium quality. But it could be same for High quality as well.

**Comment by MichaelRinger** [comment]
Ok, Ive tried your code with yolov5n and the inference time is around 10 times higher. Regarding the input size of 640x640 it still is way to high. After that I validated my implementation by swapping out the models with the ssd_mobilnet and I got an inference time of 150 ms. Since the inference code is just `_interpreter.run()` idk where the problem could be. Do you have any idea what I can try or check @VadymPinchuk?

**Comment by PaulTR** [comment]
@MichaelRinger Do you have a link for the model? Out of curiosity I'd want to run it through the [Android Object Detection sample for MediaPipe](https://github.com/googlesamples/mediapipe/tree/main/examples/object_detection/android) to see if that helps isolate it to this repo vs the model.

**Comment by fepfitra** [comment]
@MichaelRinger do you use custom model? Could you show me the way you train it?

**Comment by MichaelRinger** [comment]
@PaulTR It is the default pretrained model from the [ultralytics repository](https://github.com/ultralytics/yolov5). The model is in pytorch's format by default. I converted it via the in the repository available `export.py` script to tflite by running `python export.py --weights yolov5n.pt --include tflite`.

yolov5n.pt model weights:
https://github.com/ultralytics/yolov5/releases/download/v7.0/yolov5n.pt

**Comment by MichaelRinger** [comment]
@PaulTR, the inference of ayolov8s model also takes about 1.6 seconds, rendering it virtually unusable. Furthermore, due to the yolo models' demand for normalized pixel values, the preprocessing time is also extended.

**Comment by PaulTR** [comment]
@MichaelRinger Alright let's go ahead and move this into an independent issue that can be tracked. Model is missing metadata for other sample apps I have (tflite, mediapipe) so I'd need to put something custom together to test this model against an existing known platform baseline.

**Comment by PaulTR** [comment]
> Hey @PaulTR
> 
> 1. I cant run any apps on my work iPhone, so I am not able to test iOS version.
> 2. Regarding orientation, I would love to fix it, but till not faced camera rotation issue, which always gives 90 degrees.
>    It will require additional work for covering all the things.  (now I remembered why I was saving into a file before the detection (some time ago) because it gives ability to read rotation from exif headers)
> 
> If it is legal requirement, I will work on it during the week. Thanks

Sorry I definitely missed this response!

1. I think the iOS version is a major requirement. I don't see any border or classification UI coming up regardless of the orientation. Maybe someone else can jump in on this though?
2. I'm less worried about orientation. It's a nice to have and something we require from our official samples on MediaPipe, but this is a community project and not "hey let's make sure this is absolutely done" :P I think the image classification live stream sample that was just updated (PR is still open) might cover this FWIW.

Other than that I think this is awesome on Android! Looking forward to getting it merged in as an example for everyone.

**Comment by codscino** [comment]
I confirm on my iPhone 11 iOS 16.5.1 the rectangles are not showing. Another problems is that the camera view is not respecting the proportions, everything looks wider. Another issue is that if I try with android simulator a yolov5.tflite model the bottom stats disappear and the model does not detect anything. Thank you anyways, this look very promising, but I am a beginner and I am afraid I cannot help so much.

**Comment by VadymPinchuk** [comment]
@PaulTR  I have asked my friend to look into it :) he is iOS eng initially. So definitely have iOS device. Thanks for flagging. Will look into it

**Comment by VadymPinchuk** [comment]
Latest update on this:  WIP

Not very obvious thing, which happened in iOS is conversion of CameraImage to image.Image: convertBGRA8888ToImage - this method worked bad, and resulting image is of very poor quality. My friend is doing his best to make it right for iOS.

**Comment by VadymPinchuk** [comment]
@PaulTR  we are happy to give an update - iOS version is fixed now. Please check and give a green/red light. Thank you

**Comment by PaulTR** [comment]
Nice! So I am seeing some weirdness with the camera->scaling for the model->rescaling for display in iOS.

In portrait I'm seeing a shift down and to the right:
IMG_0130
IMG_0131

I think that's the only thing that'd be worth fixing, otherwise I think we're golden :)

**Comment by VadymPinchuk** [comment]
@PaulTR also fun thing to see - iOS prediction is faster. As everything else. Comparing to my Pixel 7.

**Comment by PaulTR** [comment]
Oh interesting, I didn't think about it with Android tablets, too. For native Android we have a scale ratio that we pull from the device and use for scaling down (to 224x224) and then back up. Using a 5th gen iPad Air running 16.5.1

**Comment by fepfitra** [comment]
@VadymPinchuk I ran example from your fork [here](https://github.com/VadymPinchuk/flutter-tflite.git) with custom model which I got [here](https://colab.research.google.com/github/EdjeElectronics/TensorFlow-Lite-Object-Detection-on-Android-and-Raspberry-Pi/blob/master/Train_TFLite2_Object_Detction_Model.ipynb#scrollTo=5tdDbTmHYwu-) but I got an error
image

I've changed the output shape by rearrange it to match to my model. But, I got negative value of bounding box.

**Comment by VadymPinchuk** [comment]
@PaulTR I have checked on iPad, changed renderBox a bit. Now it works fine on Pixel, iPhone, iPad (what I was able to check). Please verify on yours :)

**Comment by VadymPinchuk** [comment]
@PaulTR  we are ready :) when you have time

**Comment by PaulTR** [comment]
Thanks! I'll check it out Monday :)

**Comment by codscino** [comment]
@MichaelRinger any idea how to use yolov8 here? If I just change the model then I get no labels and no stats about times.

**Comment by MichaelRinger** [comment]
@codscino [Here](https://github.com/tensorflow/flutter-tflite/issues/112) is a issue that has the test code for running yolov8. Unfortunatelly its really slow with this package and not really usable for real time applications.

**Comment by codscino** [comment]
@VadymPinchuk Any idea how to integrate yolo v8 in this amazing example? Here is a working repo https://github.com/codscino/yolo_tflite, but it works only with photo in assets folder and it is pretty slow.

**Comment by VadymPinchuk** [comment]
@codscino sorry but I am not ML specialist, so from my perspective it is a black box, which takes some input and return some output :) I can spend some time with that example, but it would not be fast. While with static images almost every model should work, as we dont need any realtime update on the screen, and slow/fast models could be applied.

**Comment by codscino** [comment]
@VadymPinchuk thanks for the quick reply. I think it is more a problem of input e output formatting between ssd and yolo. I will try to investigate as well. I thought you had some insights by just seeing my super simple example. Btw thanks for this PR, it is super helpful.

**Comment by ArdeoDeo** [comment]
@codscino

1. change model name in model path
2. change image resize to 640x640
3. check if input image pixels need to be normalized or not
4. change output shape to [1,84,8400] example:
```
final output = [List<List<double>>
          .filled(84, List<double>
          .filled(8400, 0.0))],
```
5. change interpretation of the output in the post process part to get bounding boxes cause mobile_net model returns [1, 1001] shape, so it have different data

and I think it should be it

**Comment by codscino** [comment]
@ArdeoDeo thanks a lot. Tomorrow I will try. Do you think is it possible to downsize to 320x320 for performance reasons?

**Comment by ArdeoDeo** [comment]
@codscino Yolo model input is 640x640 so it need to be that size

**Comment by PaulTR** [comment]
LGTM! Thanks for getting this together. One small issue I ran into on iOS for delegates, but it can be addressed in a separate PR (I'm guessing we can just default to the CPU delegate instead of metal and call it good for an example use-case.) There's also a few errors in here about orientation and image format that can be addressed, but they're not show-stoppers. I just had to really wait for iOS to build and run.

```
2023-07-31 11:45:42.050505-0600 Runner[1579:440604] Metal GPU Frame Capture Enabled
2023-07-31 11:45:42.050768-0600 Runner[1579:440604] Metal API Validation Enabled
2023-07-31 11:45:42.269460-0600 Runner[1579:440604] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(35)] Using the Impeller rendering backend.
2023-07-31 11:45:43.165071-0600 Runner[1579:441154] flutter: The Dart VM service is listening on http://127.0.0.1:61699/y8gue7ETImM=/
2023-07-31 11:45:43.307867-0600 Runner[1579:440604] Failed to change device orientation: Error Domain=UISceneErrorDomain Code=101 "The current windowing mode does not allow for programmatic changes to interface orientation." UserInfo={NSLocalizedDescription=The current windowing mode does not allow for programmatic changes to interface orientation.}
2023-07-31 11:45:44.110545-0600 Runner[1579:441126] Created TensorFlow Lite delegate for Metal.
INFO: Created TensorFlow Lite delegate for Metal.
2023-07-31 11:45:44.117229-0600 Runner[1579:441150] The selected imageFormatGroup is not supported by iOS. Defaulting to brga8888
2023-07-31 11:45:44.131268-0600 Runner[1579:441126] Initialized TensorFlow Lite runtime.
INFO: Initialized TensorFlow Lite runtime.
2023-07-31 11:45:44.133896-0600 Runner[1579:441126] Following operations are not supported by GPU delegate:
CUSTOM TFLite_Detection_PostProcess: TFLite_Detection_PostProcess
63 operations will run on the GPU, and the remaining 1 operations will run on the CPU.
ERROR: Following operations are not supported by GPU delegate:
CUSTOM TFLite_Detection_PostProcess: TFLite_Detection_PostProcess
63 operations will run on the GPU, and the remaining 1 operations will run on the CPU.
```

**Comment by VadymPinchuk** [comment]
@PaulTR I can check these issues and make an update in a separate PR later today-tomorrow.
